module5/lessons/Week4/ProcessOptimization.md

```diff
@@ -0,0 +1,246 @@
+
+<section class="call-to-action" markdown='1'>
+
+Let's practice solving a problem at a high level.  Just focus on writing out pseudo code, **NOT** actual code.
+
+**Problem:**
+- You have an array which contains all numbers from 1 to 1 million.
+- The numbers are randomly ordered/shuffled in the array.
+- One number is in the array twice, also at some random location.
+- How might you find the duplicate value?
+
+    <section class="note" markdown='1'>
+    ### Note
+
+    There are a couple of solutions to the problem above that have various pros and cons.  Don't worry about trying to get the perfect solution.  Instead, practice breaking down the problem and thinking about the approach you could take.
+    </section>
+</section>
+
+<section class="answer" markdown='1'>
+### Potential Solution A  
+
+**Sort/Compare**
+```
+// sort numbers in ascending order
+// iterate over numbers
+  // look at neighbor.  If they are the same, return the value
+```
+
+One potential solution is to sort the array of numbers and then iterate over the numbers again, comparing each value to it's neighbor to see if they are the same.
+</section>
+
+<section class="answer" markdown='1'>
+### Potential Solution B  
+
+**Nested Lookup**
+```
+// iterate over array of numbers
+  // iterate over the array again and compare it to the rest of the array
+    // If a match is found, return the value
+```
+
+Another potential solution is to iterate over the array of numbers, and for each iteration, compare it to the rest of the array to see if a duplicate is found.
+</section>
+
+<section class="answer" markdown='1'>
+### Potential Solution C  
+
+**Hash/Obj Tracker**
+```
+// iterate over the array and add each value to a hash
+  // As soon as the value of a key is 2, you've found the duplicate
+```
+
+A third solution might include iterating over an array and adding it as a key to an object/hash, counting the number of instances in the value.  As soon as a key already exists and is increased to 2, you've found your duplicate.
+</section>
+
+### Choosing an Algorithm
+All of the options above will work to solve the problem, so how do we know which is 'best'?  We need some way of calculating or describing the time-complexity of an algorithm.  Typically, we describe time complexity in terms of how many steps it will take to complete in relation to an increasing number of elements.
+
+<section class='call-to-action' markdown='1'>
+
+In small groups, determine how many 'steps' are completed for the two code snippets below.
+* How many steps are represented by the current code?
+    * What if num[1] is a duplicate of num[0]?
+    * What if num[9] is a duplicate of num[5]?
+* How many steps would there be if `nums` had 25 elements?
+* How many steps would there be if `nums` had 50 elements?
+
+**Option 1**  
+```c#
+// find the duplicate
+int[] nums = {36, 84, 75, 20, 30, 44, 17, 98, 92, 27 };
+int? duplicate;
+
+foreach (var numA in nums)
+{
+    foreach (var numB in nums)
```

Comment:
My hunch is that this code wouldn't work because it would mark there as a duplicate when it's really just that numA and numB are the exact same number.

Reply:
I checked it in a repl and looks like yup it always gives the last number.

Not sure how to do it with a foreach, but you could do something like this

`public static void Main (string[] args) {
    int[] nums = { 36, 84, 75, 20, 30, 44, 17, 98, 44, 92, 27 };
    int? duplicate = null; // Initialize duplicate as null

    for (int i = 0; i < nums.Length; i++)
    {
        for (int j = i + 1; j < nums.Length; j++)
        {
            if (nums[i] == nums[j])
            {
                duplicate = nums[i];
            }
        }
    }
    Console.WriteLine(duplicate);
  }`

Reply:
Or looks like this world too if you want the full badness of time complexity 😆 

` int[] nums = { 36, 84, 75, 20, 30, 44, 17, 98, 44, 92, 27 };
    int? duplicate = null; // Initialize duplicate as null

    for (int i = 0; i < nums.Length; i++)
    {
        for (int j = 0; j < nums.Length; j++)
        {
            if (nums[i] == nums[j] && i != j)
            {
                duplicate = nums[i];
            }
        }
    }
    Console.WriteLine(duplicate);
  }
`

Reply:
Also Repl.it is SOOOO slow that you can really see the time differences in these solutions. Might be worth running these solutions during the lesson do really see how they have very different times to run.

Reply:
Although I think also it's doing some caching magic because it's getting quicker the more I run the program 🤷‍♀️

Reply:
Lol - this is what happens when I'm too confident in my abilities 😆 Thanks!!